MoreLinq/Experimental/Batch.cs

```diff
@@ -0,0 +1,294 @@
+        /// </typeparam>
+        /// <param name="source">The source sequence.</param>
+        /// <param name="size">Size of buckets.</param>
+        /// <param name="pool">The pool used to rent the array for each bucket.</param>
+        /// <param name="bucketProjectionSelector">A function that returns a
+        /// sequence projection to use for each bucket. It is called initially
+        /// before iterating over <paramref name="source"/>, but the resulting
+        /// projection is evaluated for each bucket.
+        /// </param>
+        /// <param name="resultSelector">A function that projects a result from
+        /// the input sequence produced over a bucket.</param>
+        /// <returns>
+        /// A sequence whose elements are projected from each bucket (returned by
+        /// <paramref name="resultSelector"/>).
+        /// </returns>
+        /// <remarks>
+        /// <para>
+        /// This operator uses deferred execution and streams its results
+        /// (buckets are streamed but their content buffered).</para>
+        /// <para>
+        /// <para>
+        /// Each bucket is backed by a rented array that may be at least
+        /// <paramref name="size"/> in length.
+        /// </para>
+        /// <para>
+        /// When more than one bucket is streamed, all buckets except the last
+        /// is guaranteed to have <paramref name="size"/> elements. The last
+        /// bucket may be smaller depending on the remaining elements in the
+        /// <paramref name="source"/> sequence.</para>
+        /// Each bucket is pre-allocated to <paramref name="size"/> elements.
+        /// If <paramref name="size"/> is set to a very large value, e.g.
+        /// <see cref="int.MaxValue"/> to effectively disable batching by just
+        /// hoping for a single bucket, then it can lead to memory exhaustion
+        /// (<see cref="OutOfMemoryException"/>).
+        /// </para>
+        /// </remarks>
+
+        public static IEnumerable<TResult>
+            Batch<TSource, TBucket, TResult>(
+                this IEnumerable<TSource> source, int size, ArrayPool<TSource> pool,
+                Func<ICurrentList<TSource>, IEnumerable<TBucket>> bucketProjectionSelector,
+                Func<IEnumerable<TBucket>, TResult> resultSelector)
+        {
+            if (source == null) throw new ArgumentNullException(nameof(source));
+            if (pool == null) throw new ArgumentNullException(nameof(pool));
+            if (size <= 0) throw new ArgumentOutOfRangeException(nameof(size));
+            if (bucketProjectionSelector == null) throw new ArgumentNullException(nameof(bucketProjectionSelector));
+            if (resultSelector == null) throw new ArgumentNullException(nameof(resultSelector));
+
+            return _(); IEnumerable<TResult> _()
+            {
+                using var batch = source.Batch(size, pool);
+                var bucket = bucketProjectionSelector(batch.CurrentList);
```

Comment:
What advantage does this additional hook point provide for potential consumers? It seems to me that any changes that will work in this context (i.e. call `bucketProjectionSelector` once, but reuse the resultant `IEnumerable<>`) can be done inside `resultSelector`.

Reply:
If you want to compute a sequence over each bucket (e.g. for sampling, filtering or what have you), then `bucketProjectionSelector` allows you to build or set-up that computation once and have it reused as each bucket is loaded. As the documentation says:

> A function that returns a sequence projection to use for each bucket. It is called initially before iterating over `source`, but the resulting projection is evaluated for each bucket.

This avoids re-allocating the same composition of iterators and closures when they don't really change with each bucket. Compare the following two:

```c#
var sample = 100;
var scale = 2.0;

var q1 =
    MoreEnumerable.Random()
                  .Take(1_000_000)
                  .Batch(10_000, ArrayPool<int>.Shared,
                         xs => xs.TakeEvery(sample).Select(x => x * scale).Average());

var q2 =
    MoreEnumerable.Random()
                  .Take(1_000_000)
                  .Batch(10_000, ArrayPool<int>.Shared,
                         xs => xs.TakeEvery(sample).Select(x => x * scale),
                         Enumerable.Average);
```

The end result is the same, but in the second case, the `xs.TakeEvery(sample).Select(x => x * scale)` bit gets initialized once but evaluated for each bucket. Consequently, the two iterators, `TakeEvery` and `Select`, and the closure `x => x * scale` get allocated once rather than for each bucket.

Reply:
Gotcha. I'm good w/ this.